### PR TITLE
Vote history

### PR DESCRIPTION
**Pull request by tbfleming**
Vote history. Also objectifies members.

### CODE REVIEW

File: contracts/eden/src/eden-micro-chain.cpp [hunk 1]
```diff
@@ -255,13 +302,145 @@ const auto& get(Table& table, const Key& key)
...
+
+VoteConnection Member::votes(std::optional<uint32_t> first,
+                             std::optional<uint32_t> last,
+                             std::optional<std::string> before,
+                             std::optional<std::string> after) const
+{
+   return clchain::make_connection<VoteConnection, vote_key>(
+       std::nullopt,                                 // gt
+       vote_key{account, 0},                         // ge
+       vote_key{eosio::name{account.value + 1}, 0},  // lt
```
Comment (swatanabe) [inline contracts/eden/src/eden-micro-chain.cpp:434]: Or use `le` and no `+1`

File: contracts/eden/src/eden-micro-chain.cpp [hunk 2]
```diff
@@ -187,22 +197,59 @@ struct member_object : public chainbase::object<member_table, member_object>
...
 };
 using member_index = mic<member_object, ordered_by_id<member_object>, ordered_by_pk<member_object>>;
 
+using ElectionGroupByRoundKey = std::tuple<uint8_t, uint64_t>;
+
+struct election_group_object : public chainbase::object<election_group_table, election_group_object>
+{
+   CHAINBASE_DEFAULT_CONSTRUCTOR(election_group_object)
+
+   // TODO: identify election
```
[the remark this answers is not in the document]
Comment (tbfleming) [reply contracts/eden/src/eden-micro-chain.cpp:210]: yes

File: contracts/eden/src/eden-micro-chain.cpp [hunk 3]
```diff
@@ -255,13 +302,145 @@ const auto& get(Table& table, const Key& key)
...
+   uint32_t auctionDuration() const { return status->auctionDuration; }
+   const std::string& memo() const { return status->memo; }
+};
+EOSIO_REFLECT2(Status, active, community, initialMembers, genesisVideo, auctionDuration, memo)
+
+struct ElectionGroup
+{
+   const election_group_object* obj;
+
+   uint64_t id() const { return obj->id._id; }
```
[the remark this answers is not in the document]
Comment (tbfleming) [reply contracts/eden/src/eden-micro-chain.cpp:397]: I suspect the UI may need some way to uniquely identify groups

File: contracts/eden/src/eden-micro-chain.cpp (hunk 1 above)
Comment (tbfleming) [reply contracts/eden/src/eden-micro-chain.cpp]: changed

File: contracts/eden/src/eden-micro-chain.cpp [hunk 4]
```diff
@@ -390,16 +407,68 @@ struct Status
...
+                                                std::optional<uint32_t> first,
+                                                std::optional<uint32_t> last,
+                                                std::optional<std::string> before,
+                                                std::optional<std::string> after) const
+{
+   return clchain::make_connection<ElectionGroupConnection, ElectionGroupByRoundKey>(
+       gt ? std::optional{ElectionGroupByRoundKey{obj->time, *gt, ~uint64_t(0)}}           //
+          : std::nullopt,                                                                  //
+       ge ? std::optional{ElectionGroupByRoundKey{obj->time, *ge, 0}}                      //
+          : std::optional{ElectionGroupByRoundKey{obj->time, 0, 0}},                       //
```
Comment (swatanabe) [inline contracts/eden/src/eden-micro-chain.cpp:507]: This relies on the fact that `gt` has higher priority that `ge`, which is not exactly obvious.

**Review by swatanabe — APPROVED** [review]

**Pull request merged**